Describe workaround for long checkout times

dscho opened the pull request:
When I tried this in https://github.com/dscho/git-sdk-64, the `checkout` step took a whopping five minutes. We don't need the files, and we don't even need to transfer all those Git objects: partial clone is here to save the day.

It looks a bit more complicated than the originally-proposed solution, so I describe this only as an alternative instead of replacing the short-and-sweet solution using `actions/checkout@v2`.

dscho committed:
Describe workaround for long checkout times

When I tried this in https://github.com/dscho/git-sdk-64, the `checkout`
step took a whopping five minutes. We don't need the files, and we don't
even need to transfer all those Git objects: partial clone is here to
save the day.

It looks a bit more complicated than the originally-proposed solution,
so I describe this only as an alternative instead of replacing the
short-and-sweet solution using `actions/checkout@v2`.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

chancancode commented:
@dscho Thanks for pointing this out. When I was writing/testing this I was a bit lazy and didn't fully dig into what is the actual minimal solution needed here. The approach looks good to me (and I agree it's good to make it a follow-up subsection), but I suspect we can simplify it a bit.

I would like to investigate whether it is possible to use the [options for the checkout actions](https://github.com/actions/checkout#usage) to accomplish the same thing.

If not, I think it would be nicer and easier to continue using `actions/checkout@v2` for checking out the repo and configuring the token, without specifying `fetch-depth: 0`. By default it only fetch a single commit which should be quite fast, and it would set up the repo, with the origin/token set, etc. Then, we can follow up with another step to do the actual fetching with a `git fetch` command. That seems to be in the spirit of how they recommend the action to be used to do custom things: https://github.com/actions/checkout#scenarios

What do you think?

As far as the actual fetching, is the idea here here basically that just specify a large enough depth (100 in your example) and hope that it is enough? (I agree it should be fine in practice.)

dscho commented:
> I would like to investigate whether it is possible to use the [options for the checkout actions](https://github.com/actions/checkout#usage) to accomplish the same thing.

As far as I can see, there is no way to specify a partial clone (I suspect it is in the making, though).

> By default it only fetch a single commit which should be quite fast, and it would set up the repo, with the origin/token set

In my case, that would not change much, as the Git for Windows SDK needs to lay down 154,008 files, and the fetch would still transfer several hundreds of megabytes.



> As far as the actual fetching, is the idea here here basically that just specify a large enough depth (100 in your example) and hope that it is enough?

Now that you got me thinking about this: even a depth of 1 should suffice, as one of both branches is already at the correct revision, i.e. the remote has all objects to begin with.

dscho commented:
> > As far as the actual fetching, is the idea here here basically that just specify a large enough depth (100 in your example) and hope that it is enough?
> 
> Now that you got me thinking about this: even a depth of 1 should suffice, as one of both branches is already at the correct revision, i.e. the remote has all objects to begin with.

Fixed.

chancancode commented on README.md:
````diff
@@ -223,6 +223,21 @@ practice, one of the two branches (the one that was pushed to) already has the
+the steps with this:
+
+```yaml
+      - name: Partial clone
+        run: git clone --bare --depth=1 --single-branch --filter=blob:none ${{ github.event.repository.html_url }} .
+      - name: Push
+        run: |
+          git config http.https://github.com/.extraheader "Authorization: Basic $(echo -n x-access-token:${{ github.token }} | base64 --wrap=0)"
+          git push origin HEAD:master HEAD:main
````
```suggestion
- name: Configure Git
  run: git config http.https://github.com/.extraheader "Authorization: Basic $(echo -n x-access-token:${{ github.token }} | base64 --wrap=0)"
- name: Partial Clone
  run: git clone --bare --depth=1 --single-branch --filter=blob:none ${{ github.event.repository.html_url }} .
- name: Push
  run: git push origin HEAD:master HEAD:main
```

Small suggestion for readability (I think it does the same thing?)

This is cool, I didn't realize you can push to multiple branches at once. If it's convenient for you, can you update the main workflow to do the same as well? If not, no worries – just commit or discard my suggestion from the UI and I can update the main one later.

Can you confirm that `--depth=1` (and `--single-branch`) will work when the push contains multiple commits? 

Admittedly my knowledge here is on very shaky ground, but I copied `fetch-depth: 0` from [an actual workflow I have in another repo](https://github.com/skylightio/skylight-docs/blob/master/.github/workflows/sync.yml). In that case, I am trying to mirror two identical repositories (an internal/private version, and the public one I linked). After some trial-and-error, I determined that `fetch-depth: 0` works.

At the time, my mental model was that I need to fetch all the commits involved the push + at least the parent of the last commit in the push, so that Git can see that the histories on the two repos are related. Otherwise, it will refuse to do the push without `--force` due to unrelated history. I would have guessed that this applies here as well, so I am a bit surprised that this (and `--single-branch`) worked.

Evidently, my mental model is not entirely correct here, so as long as you can confirm that it works then it's all good.

dscho replied:
> Small suggestion for readability (I think it does the same thing?)

Not quite: that `git config` call will fail because there is no Git repository (the `git clone` command will create that). I re-ordered it and force-pushed.

> Can you confirm that `--depth=1` (and `--single-branch`) will work when the push contains multiple commits?

I am 100% certain, even without testing it.

> Admittedly my knowledge here is on very shaky ground, but I copied `fetch-depth: 0` from [an actual workflow I have in another repo](https://github.com/skylightio/skylight-docs/blob/master/.github/workflows/sync.yml). In that case, I am trying to mirror two identical repositories (an internal/private version, and the public one I linked). After some trial-and-error, I determined that `fetch-depth: 0` works.

Those are mirrors between different repositories, right? In that case, Git objects will need to be pushed because they are unlikely to be present in the target repository yet. However, here we are operating with the same remote repository. We know it has all the objects because that's where we fetch from.

> At the time, my mental model was that I need to fetch all the commits involved the push + at least the parent of the last commit in the push, so that Git can see that the histories on the two repos are related. Otherwise, it will refuse to do the push without `--force` due to unrelated history. I would have guessed that this applies here as well, so I am a bit surprised that this (and `--single-branch`) worked.

Git needs to know the hash of the parent commit in the case that you want to push new objects to a repository, indeed.

But again, this does not apply here. We're trying to update a remote ref to a commit that we know is already there. You can actually see this in the output when pushing manually (in CI, the progress is suppressed because there is no interactive terminal):

```
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
[...]
    * [new branch]     HEAD:main
```

You see, it does not transfer any objects. It does not need to.

(Technical explanation: to save on bandwidth, Git determines the delta between the tip commit(s) to push and the tip commits reported by the remote repository, of _all_ refs. In other words, Git will see that the remote repository does indeed have all the objects already and make an empty pack.)

chancancode replied:
@dscho just woke up and still trying to digest all of this - but the burning question I have is: does that mean fetch-depth: 0 is also not necessary in the main workflow? 😅 I should probably have checked at the time

chancancode replied:
(and if that’s the case feel free to update the script and delete the explanation text for me, or I can do it later; either way this looks good and I will merge when I can computer)

dscho replied:
> does that mean fetch-depth: 0 is also not necessary in the main workflow?

You will be delighted to hear that it is not ;-)

> and if that’s the case feel free to update the script and delete the explanation text for me, or I can do it later; either way this looks good and I will merge when I can computer

Whoops, I think I only updated the script.

*clicketyclick*

Yep, I forgot to update the descriptions! And I also missed a few more mentions of `fetch-depth: 0`. Sorry about that. I force-pushed a fixed version (grepping for the term `fetch-depth` to make sure that I did not miss any instance).

chancancode replied:
@dscho This makes a lot of sense to me now, thank you so much for helping me understand this and your help with updating the guide!! 🎉

dscho replied:
Awesome! Thank you for starting this very helpful guide. I used it to plan Git for Windows' repositories' transition.

dscho committed:
Simplify the mirror action

You can push multiple refs in the same `push`.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Make the mirror workflow more efficient

No need to fetch the entire history: all of the objects that are
necessary for the subsequent push are already on the remote side.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Small grammar fix

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
